dev/archery/archery/utils/command.py

```diff
@@ -0,0 +1,67 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+import os
+import shutil
+import subprocess
+
+from .logger import logger, ctx
+
+
+def find_exec(executable):
+    exec_exists = os.path.exists(executable)
+    return executable if exec_exists else shutil.which(executable)
+
+
+# Decorator running a command and returning stdout
+class capture_stdout:
+    def __init__(self, strip=False):
+        self.strip = strip
+
+    def __call__(self, f):
+        def strip_it(x):
+            return x.strip() if self.strip else x
+
+        def wrapper(*argv, **kwargs):
+            # Ensure stdout is captured
+            kwargs["stdout"] = subprocess.PIPE
+            return strip_it(fn(*argv, **kwargs).stdout)
+        return wrapper
+
+
+class Command:
+    def bin(self):
```

Comment:
elsewhere you seem to be using this as a property

Reply:
I'm not sure what's the best way to explicit an interface.

Reply:
Perhaps [`abc`](https://docs.python.org/3/library/abc.html)? You can make an abstract property.

Reply:
Would it be acceptable to put the requirement in a doc string, remove this method, and add a check for `self.bin` to `run()`?

```
    def run(self, *argv, **kwargs):
        if not hasattr(self, 'bin'):
            raise NotImplementedError('Command must provide bin property')
```

Reply:
I'll got with the hasattr, since this is not user facing but developer facing. It'll bork loudly only once.